src/Microsoft.ML.Data/DataView/InternalSchemaDefinition.cs

```diff
@@ -185,9 +186,13 @@ public static void GetVectorAndItemType(Type rawType, string name, out bool isVe
                 isVector = false;
             }
 
+            // The internal type of string is ReadOnlyMemory<char>. That is, string will be stored as ReadOnlyMemory<char> in IDataView.
             if (itemType == typeof(string))
                 itemType = typeof(ReadOnlyMemory<char>);
-            else if (!itemType.TryGetDataKind(out _))
+
+            // Check if the itemType extracted from rawType is supported by ML.NET's type system.
+            // It must be one of either ML.NET's pre-defined types or custom types registered by the user.
+            if (!itemType.TryGetDataKind(out _) && !DataViewTypeManager.Knows(itemType, attributes))
```

Comment:
Why did you remove the `else` here? We don't need to call this for `string`.

Reply:
The semantic here was not quite consistent. The string branch is to replace a type. However the `else if` was for checking if the type is allowed.

---
In reply to: [287032519](https://github.com/dotnet/machinelearning/pull/3263#discussion_r287032519) [](ancestors = 287032519)

Reply:
I don't understand your response. Why would we need to check if `string` is allowed?  Of course it is. If `itemType == typeof(string)`, replace it to `ROM<char>` and no need to check if it if valid, because we know it is valid.

---
In reply to: [287136944](https://github.com/dotnet/machinelearning/pull/3263#discussion_r287136944) [](ancestors = 287136944,287032519)

Reply:
That's fine. I will change it back.

---
In reply to: [287165481](https://github.com/dotnet/machinelearning/pull/3263#discussion_r287165481) [](ancestors = 287165481,287136944,287032519)